ports/stm32/mphciport.c

```diff
@@ -135,60 +123,83 @@ int mp_bluetooth_hci_uart_readchar(void) {
-    mp_bluetooth_hci_uart_obj.timeout = 2;
-    mp_bluetooth_hci_uart_obj.timeout_char = 2;
+    // We don't want to block indefinitely, but expect flow control is doing its job.
+    mp_bluetooth_hci_uart_obj.timeout = 200;
+    mp_bluetooth_hci_uart_obj.timeout_char = 200;
     MP_STATE_PORT(pyb_uart_obj_all)[mp_bluetooth_hci_uart_obj.uart_id - 1] = &mp_bluetooth_hci_uart_obj;
+
+    // This also initialises the UART.
+    mp_bluetooth_hci_uart_set_baudrate(baudrate);
+
     return 0;
 }
 
-int mp_bluetooth_hci_uart_set_baudrate(uint32_t baudrate) {
-    uart_init(&mp_bluetooth_hci_uart_obj, baudrate, UART_WORDLENGTH_8B, UART_PARITY_NONE, UART_STOPBITS_1, UART_HWCONTROL_RTS | UART_HWCONTROL_CTS);
-    uart_set_rxbuf(&mp_bluetooth_hci_uart_obj, sizeof(hci_uart_rxbuf), hci_uart_rxbuf);
+int mp_bluetooth_hci_uart_deinit(void) {
+    DEBUG_printf("mp_bluetooth_hci_uart_deinit (stm32)\n");
+    // TODO: deinit mp_bluetooth_hci_uart_obj
+
     return 0;
 }
 
-int mp_bluetooth_hci_uart_activate(void) {
-    // Interrupt on RX chunk received (idle)
-    // Trigger stack poll when this happens
-    mp_obj_t uart_irq_fn = mp_load_attr(MP_OBJ_FROM_PTR(&mp_bluetooth_hci_uart_obj), MP_QSTR_irq);
-    mp_obj_t uargs[] = {
-        MP_OBJ_FROM_PTR(&mp_uart_interrupt_obj),
-        MP_OBJ_NEW_SMALL_INT(UART_FLAG_IDLE),
-        mp_const_true,
-    };
-    mp_call_function_n_kw(uart_irq_fn, 3, 0, uargs);
+int mp_bluetooth_hci_uart_set_baudrate(uint32_t baudrate) {
+    DEBUG_printf("mp_bluetooth_hci_uart_set_baudrate(%lu) (stm32)\n", baudrate);
+    if (!baudrate) {
+        return -1;
+    }
+
+    uart_init(&mp_bluetooth_hci_uart_obj, baudrate, UART_WORDLENGTH_8B, UART_PARITY_NONE, UART_STOPBITS_1, UART_HWCONTROL_RTS | UART_HWCONTROL_CTS);
+    uart_set_rxbuf(&mp_bluetooth_hci_uart_obj, sizeof(hci_uart_rxbuf), hci_uart_rxbuf);
 
-    mp_bluetooth_hci_controller_init();
-    mp_bluetooth_hci_controller_activate();
+    // Add IRQ handler for IDLE (i.e. packet finished).
+    uart_irq_config(&mp_bluetooth_hci_uart_obj, false);
+    mp_irq_init(&mp_bluetooth_hci_uart_irq_obj, &uart_irq_methods, MP_OBJ_FROM_PTR(&mp_bluetooth_hci_uart_obj));
+    mp_bluetooth_hci_uart_obj.mp_irq_obj = &mp_bluetooth_hci_uart_irq_obj;
+    mp_bluetooth_hci_uart_obj.mp_irq_trigger = UART_FLAG_IDLE;
+    mp_bluetooth_hci_uart_irq_obj.handler = MP_OBJ_FROM_PTR(&mp_uart_interrupt_obj);
+    mp_bluetooth_hci_uart_irq_obj.ishard = true;
+    uart_irq_config(&mp_bluetooth_hci_uart_obj, true);
 
     return 0;
 }
 
 int mp_bluetooth_hci_uart_write(const uint8_t *buf, size_t len) {
+    // DEBUG_printf("mp_bluetooth_hci_uart_write (stm32)\n");
+
     mp_bluetooth_hci_controller_wakeup();
-    uart_tx_strn(&mp_bluetooth_hci_uart_obj, (void *)buf, len);
+    int errcode;
+    uart_tx_data(&mp_bluetooth_hci_uart_obj, (void *)buf, len, &errcode);
+    if (errcode != 0) {
+        printf("\nmp_bluetooth_hci_uart_write: failed to write to UART %d\n", errcode);
+    }
```

Comment:
In my branch I've changed this if to `assert(errcode == 0);`

My updated commit is here, it includes your comment and removes the extra comment/printf below:
https://github.com/andrewleech/micropython/commit/32bf8073d9dc843abc95aca48ed369ab854db8f6

Reply:
I think this is too important to be a debug-only notification...

Reply:
Ah yeah, I forgot that asserts are disabled in non-debug builds... yep perhaps a static flag that can be checked at a higher layer will be needed, perhaps something that can then trigger a stack/hci reset.